File: DDCollectionViewFlowLayout/Classes/DDCollectionViewFlowLayout.m
//
//  DDCollectionViewFlowLayout.m
//  DDCollectionViewFlowLayout
//
//  Created by DeJohn Dong on 15-2-12.
//  Copyright (c) 2015年 DDKit. All rights reserved.
//

#import "DDCollectionViewFlowLayout.h"

@interface DDCollectionViewFlowLayout()

@property (nonatomic, strong) NSMutableArray *sectionRects;
@property (nonatomic, strong) NSMutableArray *columnRectsInSection;
@property (nonatomic, strong) NSMutableArray *layoutItemAttributes;
@property (nonatomic, strong) NSDictionary *headerFooterItemAttributes;
@property (nonatomic, strong) NSMutableArray *sectionInsetses;
@property (nonatomic) UIEdgeInsets currentEdgeInsets;

@end

@implementation DDCollectionViewFlowLayout

#pragma mark - UISubclassingHooks Category Methods

- (CGSize)collectionViewContentSize {
    [super collectionViewContentSize];
    
    CGRect lastSectionRect = [[self.sectionRects lastObject] CGRectValue];
    CGSize lastSize = CGSizeMake(CGRectGetWidth(self.collectionView.bounds), CGRectGetMaxY(lastSectionRect));
    return lastSize;
}

- (void)prepareLayout {
    
    NSUInteger numberOfSections = self.collectionView.numberOfSections;
    self.sectionRects = [[NSMutableArray alloc] initWithCapacity:numberOfSections];
    self.columnRectsInSection = [[NSMutableArray alloc] initWithCapacity:numberOfSections];
    self.layoutItemAttributes = [[NSMutableArray alloc] initWithCapacity:numberOfSections];
    self.sectionInsetses = [[NSMutableArray alloc] initWithCapacity:numberOfSections];
    self.headerFooterItemAttributes = @{UICollectionElementKindSectionHeader:[NSMutableArray array], UICollectionElementKindSectionFooter:[NSMutableArray array]};
    
    for (NSUInteger section = 0; section < numberOfSections; ++section) {
        NSUInteger itemsInSection = [self.collectionView numberOfItemsInSection:section];
        [self.layoutItemAttributes addObject:[NSMutableArray array]];
        [self prepareSectionLayout:section withNumberOfItems:itemsInSection];
    }
}

- (UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)kind
                                                                     atIndexPath:(NSIndexPath *)indexPath {
    return self.headerFooterItemAttributes[kind][indexPath.section];
}

- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
    return self.layoutItemAttributes[indexPath.section][indexPath.item];
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    return [self searchVisibleLayoutAttributesInRect:rect];
}

- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds {
    return YES;
}

#pragma mark - Private Methods

- (void)prepareSectionLayout:(NSUInteger)section withNumberOfItems:(NSUInteger)numberOfItems {
    UICollectionView *cView = self.collectionView;
    
    NSIndexPath *indexPath = [NSIndexPath indexPathForItem:0 inSection:section];
    
    //# hanlde the section header
    CGFloat headerHeight = 0.0f;
    
    CGRect previousSectionRect = [self rectForSectionAtIndex:indexPath.section - 1];
    CGRect sectionRect;
    sectionRect.origin.x = 0;
    sectionRect.origin.y = CGRectGetHeight(previousSectionRect) + CGRectGetMinY(previousSectionRect);
    sectionRect.size.width = cView.bounds.size.width;
    
    if([self.delegate respondsToSelector:@selector(collectionView:layout:referenceSizeForHeaderInSection:)]) {
        
        //# Define the rect of the header
        CGRect headerFrame;
        headerFrame.origin.x = 0.0f;
        headerFrame.origin.y = sectionRect.origin.y;

        CGSize headerSize = [self.delegate collectionView:self.collectionView layout:self referenceSizeForHeaderInSection:indexPath.section];
        headerFrame.size.height = headerSize.height;
        headerFrame.size.width = headerSize.width;
        
        UICollectionViewLayoutAttributes *headerAttributes =
        [UICollectionViewLayoutAttributes
         layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionHeader
         withIndexPath:indexPath];
        headerAttributes.frame = headerFrame;
        
        headerHeight = headerFrame.size.height;
        [self.headerFooterItemAttributes[UICollectionElementKindSectionHeader] addObject:headerAttributes];
    }

    //# get the insets of section
    UIEdgeInsets sectionInset = UIEdgeInsetsZero;
    
    if([self.delegate respondsToSelector:@selector(collectionView:layout:insetForSectionAtIndex:)]) {
        sectionInset = [self.delegate collectionView:cView layout:self insetForSectionAtIndex:section];
    }
    
    [self.sectionInsetses addObject:[NSValue valueWithUIEdgeInsets:sectionInset]];
    
    CGRect itemsContentRect;
    
    //# the the lineSpacing & interitemSpacing default values
    CGFloat lineSpacing = 0.0f;
    CGFloat interitemSpacing = 0.0f;

    if([self.delegate respondsToSelector:@selector(collectionView:layout:minimumInteritemSpacingForSectionAtIndex:)]) {
        interitemSpacing = [self.delegate collectionView:cView layout:self minimumInteritemSpacingForSectionAtIndex:section];
    }
    
    if([self.delegate respondsToSelector:@selector(collectionView:layout:minimumLineSpacingForSectionAtIndex:)]) {
        lineSpacing = [self.delegate collectionView:cView layout:self minimumLineSpacingForSectionAtIndex:section];
    }
    
    itemsContentRect.origin.x = sectionInset.left;
    itemsContentRect.origin.y = headerHeight + sectionInset.top;
    
    NSUInteger numberOfColumns = [self.delegate collectionView:cView layout:self numberOfColumnsInSection:section];
    itemsContentRect.size.width = CGRectGetWidth(cView.frame) - (sectionInset.left + sectionInset.right);
    
    CGFloat columnSpace = itemsContentRect.size.width - (interitemSpacing * (numberOfColumns - 1));
    CGFloat columnWidth = (columnSpace/numberOfColumns);
    
    // # store space for each column
    [self.columnRectsInSection addObject:[NSMutableArray arrayWithCapacity:numberOfColumns]];
    for (NSUInteger colIdx = 0; colIdx < numberOfColumns; ++colIdx)
        [self.columnRectsInSection[section] addObject:[NSMutableArray array]];
    
    // # Define the rect of the of each item
    for (NSInteger itemIdx = 0; itemIdx < numberOfItems; ++itemIdx) {
        NSIndexPath *itemPath = [NSIndexPath indexPathForItem:itemIdx inSection:section];
        CGSize itemSize = [self.delegate collectionView:cView layout:self sizeForItemAtIndexPath:itemPath];
        
        NSInteger destColumnIdx = [self preferredColumnIndexInSection:section];
        NSInteger destRowInColumn = [self numberOfItemsInColumn:destColumnIdx ofSection:section];
        CGFloat lastItemInColumnOffset = [self lastItemOffsetInColumn:destColumnIdx inSection:section];
        
        if(destRowInColumn == 0) {
            lastItemInColumnOffset += sectionRect.origin.y;
        }
        
        CGRect itemRect;
        itemRect.origin.x = itemsContentRect.origin.x + destColumnIdx * (interitemSpacing + columnWidth);
        itemRect.origin.y = lastItemInColumnOffset + (destRowInColumn > 0 ? lineSpacing: sectionInset.top);
        itemRect.size.width = columnWidth;
        itemRect.size.height = itemSize.height;
                
        UICollectionViewLayoutAttributes *itemAttributes = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:itemPath];
        itemAttributes.frame = itemRect;
        [self.layoutItemAttributes[section] addObject:itemAttributes];
        [self.columnRectsInSection[section][destColumnIdx] addObject:[NSValue valueWithCGRect:itemRect]];
    }
    
    itemsContentRect.size.height = [self heightOfItemsInSection:indexPath.section] + sectionInset.bottom;
    
    // # define the section footer
    CGFloat footerHeight = 0.0f;
    if([self.delegate respondsToSelector:@selector(collectionView:layout:referenceSizeForFooterInSection:)]) {
        CGRect footerFrame;
        footerFrame.origin.x = 0;
        footerFrame.origin.y = itemsContentRect.size.height;
        
        CGSize footerSize = [self.delegate collectionView:self.collectionView layout:self referenceSizeForFooterInSection:indexPath.section];
        footerFrame.size.height = footerSize.height;
        footerFrame.size.width = footerSize.width;
        
        UICollectionViewLayoutAttributes *footerAttributes = [UICollectionViewLayoutAttributes
                                                              layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionFooter
                                                              withIndexPath:indexPath];
        footerAttributes.frame = footerFrame;
        
        footerHeight = footerFrame.size.height;
        
        [self.headerFooterItemAttributes[UICollectionElementKindSectionFooter] addObject:footerAttributes];
    }
    
    if(section > 0){
        itemsContentRect.size.height -= sectionRect.origin.y;
    }
    
    sectionRect.size.height = itemsContentRect.size.height + footerHeight;

    [self.sectionRects addObject:[NSValue valueWithCGRect:sectionRect]];
}

- (CGFloat)heightOfItemsInSection:(NSUInteger)sectionIdx {
    CGFloat maxHeightBetweenColumns = 0.0f;
    NSArray *columnsInSection = self.columnRectsInSection[sectionIdx];
    for (NSUInteger columnIdx = 0; columnIdx < columnsInSection.count; ++columnIdx) {
        CGFloat heightOfColumn = [self lastItemOffsetInColumn:columnIdx inSection:sectionIdx];
        maxHeightBetweenColumns = MAX(maxHeightBetweenColumns, heightOfColumn);
    }
    return maxHeightBetweenColumns;
}

- (NSInteger)numberOfItemsInColumn:(NSInteger)columnIdx ofSection:(NSInteger)sectionIdx {
    return [self.columnRectsInSection[sectionIdx][columnIdx] count];
}

- (CGFloat)lastItemOffsetInColumn:(NSInteger)columnIdx inSection:(NSInteger)sectionIdx {
    NSArray *itemsInColumn = self.columnRectsInSection[sectionIdx][columnIdx];
    if (itemsInColumn.count == 0) {
        if([self.headerFooterItemAttributes[UICollectionElementKindSectionHeader] count] > sectionIdx) {
            CGRect headerFrame = [self.headerFooterItemAttributes[UICollectionElementKindSectionHeader][sectionIdx] frame];
            return headerFrame.size.height;
        }
        return 0.0f;
    } else {
        CGRect lastItemRect = [[itemsInColumn lastObject] CGRectValue];
        return CGRectGetMaxY(lastItemRect);
    }
}

- (NSInteger)preferredColumnIndexInSection:(NSInteger)sectionIdx {
    NSUInteger shortestColumnIdx = 0;
    CGFloat heightOfShortestColumn = CGFLOAT_MAX;
    for (NSUInteger columnIdx = 0; columnIdx < [self.columnRectsInSection[sectionIdx] count]; ++columnIdx) {
        CGFloat columnHeight = [self lastItemOffsetInColumn:columnIdx inSection:sectionIdx];
        if (columnHeight < heightOfShortestColumn) {
            shortestColumnIdx = columnIdx;
            heightOfShortestColumn = columnHeight;
        }
    }
    return shortestColumnIdx;
}

- (CGRect)rectForSectionAtIndex:(NSInteger)sectionIdx {
    if (sectionIdx < 0 || sectionIdx >= self.sectionRects.count)
        return CGRectZero;
    return [self.sectionRects[sectionIdx] CGRectValue];
}

#pragma mark - Show Attributes Methods
- (NSArray *)searchVisibleLayoutAttributesInRect:(CGRect)rect {
    NSMutableArray *itemAttrs = [[NSMutableArray alloc] init];
    NSIndexSet *visibleSections = [self sectionIndexesInRect:rect];
    [visibleSections enumerateIndexesUsingBlock:^(NSUInteger sectionIdx, BOOL *stop) {
        
        //# items
        for (UICollectionViewLayoutAttributes *itemAttr in self.layoutItemAttributes[sectionIdx]) {
            CGRect itemRect = itemAttr.frame;
            itemAttr.zIndex = 1;
            BOOL isVisible = CGRectIntersectsRect(rect, itemRect);
            if (isVisible)
                [itemAttrs addObject:itemAttr];
        }
        
        //# footer
        if([self.headerFooterItemAttributes[UICollectionElementKindSectionFooter] count] > sectionIdx) {
            UICollectionViewLayoutAttributes *footerAttribute = self.headerFooterItemAttributes[UICollectionElementKindSectionFooter][sectionIdx];
            BOOL isVisible = CGRectIntersectsRect(rect, footerAttribute.frame);
            if (isVisible && footerAttribute)
                [itemAttrs addObject:footerAttribute];
            self.currentEdgeInsets = UIEdgeInsetsZero;
        } else {
            self.currentEdgeInsets = [self.sectionInsetses[sectionIdx] UIEdgeInsetsValue];
        }
        
        //# header
        if([self.headerFooterItemAttributes[UICollectionElementKindSectionHeader] count] > sectionIdx) {
            UICollectionViewLayoutAttributes *headerAttribute = self.headerFooterItemAttributes[UICollectionElementKindSectionHeader][sectionIdx];
            
            if(!self.enableStickyHeaders) {
                BOOL isVisibleHeader = CGRectIntersectsRect(rect, headerAttribute.frame);
                
                if (isVisibleHeader && headerAttribute)
                    [itemAttrs addObject:headerAttribute];
            } else {
                UICollectionViewLayoutAttributes *lastCell = [itemAttrs lastObject];
                
                if(headerAttribute)
                   [itemAttrs addObject:headerAttribute];
                
                [self updateHeaderAttributes:headerAttribute lastCellAttributes:lastCell];
            }
        }
    }];
    return itemAttrs;
}

- (NSIndexSet *)sectionIndexesInRect:(CGRect)rect {
    CGRect theRect = rect;
    NSMutableIndexSet *visibleIndexes = [[NSMutableIndexSet alloc] init];
    NSUInteger numberOfSections = self.collectionView.numberOfSections;
    for (NSUInteger sectionIdx = 0; sectionIdx < numberOfSections; ++sectionIdx) {
        CGRect sectionRect = [self.sectionRects[sectionIdx] CGRectValue];
        BOOL isVisible = CGRectIntersectsRect(theRect, sectionRect);
        if (isVisible)
            [visibleIndexes addIndex:sectionIdx];
    }
    return visibleIndexes;
}

#pragma mark - Sticky Header implementation methods

- (void)updateHeaderAttributes:(UICollectionViewLayoutAttributes *)attributes
            lastCellAttributes:(UICollectionViewLayoutAttributes *)lastCellAttributes
{
    CGRect currentBounds = self.collectionView.bounds;
    attributes.zIndex = 1024;
    attributes.hidden = NO;
    
    CGPoint origin = attributes.frame.origin;
    
    CGFloat sectionMaxY = CGRectGetMaxY(lastCellAttributes.frame) - attributes.frame.size.height + self.currentEdgeInsets.bottom;
    
    CGFloat y = CGRectGetMaxY(currentBounds) - currentBounds.size.height + self.collectionView.contentInset.top;
    
    CGFloat maxY = MIN(MAX(y, attributes.frame.origin.y), sectionMaxY);
    
    origin.y = maxY;
    
    attributes.frame = (CGRect){
        origin,
        attributes.frame.size
    };
}

@end

File: Swift/DDCollectionViewFlowLayout_Swift/Classes/DDCollectionViewFlowLayout.swift
//
//  DDCollectionViewFlowLayout.swift
//  DDCollectionViewFlowLayout_Swift
//
//  Created by DeJohn Dong on 15/9/30.
//  Copyright © 2015年 ddkit. All rights reserved.
//

import UIKit

@objc protocol DDCollectionViewDelegateFlowLayout : UICollectionViewDelegateFlowLayout {
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: DDCollectionViewFlowLayout, numberOfColumnInSection section: Int) -> NSInteger
    
    
    @objc optional func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: DDCollectionViewFlowLayout, backgroundColorForSectionAtIndex index: Int) -> UIColor
}

class DDCollectionViewFlowLayout : UICollectionViewFlowLayout {
    fileprivate var sectionRects = NSMutableArray()
    fileprivate var columnRectsInSection = NSMutableArray()
    fileprivate var layoutItemAttributes = NSMutableArray()
    fileprivate var headerItemAttributes = NSMutableArray()
    fileprivate var footerItemAttributes = NSMutableArray()
    fileprivate var sectionInsetses = NSMutableArray()
    fileprivate var backgroundColorAttributes = NSMutableArray()
    fileprivate var currentEdgeInsets: UIEdgeInsets = UIEdgeInsets.zero
    
    @IBOutlet weak var delegate : DDCollectionViewDelegateFlowLayout?
    @IBInspectable var enableStickyHeaders = false
    
    override var collectionViewContentSize : CGSize {
        var lastSize = super.collectionViewContentSize
        
        let lastSectionRect = (sectionRects.lastObject! as AnyObject).cgRectValue
        lastSize = CGSize(width: (self.collectionView?.bounds.width)!, height: (lastSectionRect?.maxY)!)
        return lastSize
    }
    
    override func prepare() -> Void {
        let numberOfSections = self.collectionView!.numberOfSections
        self.layoutItemAttributes.removeAllObjects()
        self.footerItemAttributes.removeAllObjects()
        self.headerItemAttributes.removeAllObjects()
        self.sectionRects.removeAllObjects()
        self.backgroundColorAttributes.removeAllObjects()
        self.columnRectsInSection.removeAllObjects()
        for i in 0 ..< numberOfSections {
            let itemsInSection = self.collectionView?.numberOfItems(inSection: i)
            self.layoutItemAttributes.add(NSMutableArray.init(capacity: 0))
            self.prepareLayoutInSection(i, numberOfItems: itemsInSection!)
        }
    }
    
    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        return (layoutItemAttributes.object(at: indexPath.section) as AnyObject).object(at: indexPath.item) as? UICollectionViewLayoutAttributes
    }
    
    override func layoutAttributesForSupplementaryView(ofKind elementKind: String, at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        if elementKind == UICollectionElementKindSectionHeader {
            return headerItemAttributes.object(at: indexPath.section) as? UICollectionViewLayoutAttributes
        } else {
            return footerItemAttributes.object(at: indexPath.section) as? UICollectionViewLayoutAttributes
        }
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        return self.searchVisibleLayoutAttributesInRect(rect)
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return self.enableStickyHeaders
    }
    
    fileprivate func prepareLayoutInSection(_ index: Int, numberOfItems items: Int) {
        let collectionView = self.collectionView
        
        let indexPath = IndexPath.init(item: 0, section: index)
        
        let previouseSectionRect = self.rectForSectionAtIndex(indexPath.section - 1)
        
        var sectionRect: CGRect = CGRect.zero
        sectionRect.origin.x = 0
        sectionRect.origin.y = previouseSectionRect.height + previouseSectionRect.minY
        sectionRect.size.width = (collectionView?.bounds.size.width)!
        
        var headerHeight: CGFloat = 0.0
        
        if ((self.delegate?.collectionView?(self.collectionView!, layout: self, referenceSizeForHeaderInSection: index)) != nil) {
            var headerFrame: CGRect = CGRect.zero
            headerFrame.origin.x = 0.0
            headerFrame.origin.y = sectionRect.origin.y
            
            let headerSize = self.delegate?.collectionView!(self.collectionView!, layout: self, referenceSizeForHeaderInSection: index)
            
            headerFrame.size.width = (headerSize?.width)!
            headerFrame.size.height = (headerSize?.height)!
            
            let headerAttributes = UICollectionViewLayoutAttributes.init(forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, with: indexPath)
            
            headerAttributes.frame = headerFrame
            
            headerHeight = headerFrame.size.height
            
            self.headerItemAttributes.add(headerAttributes)
        }
        
        var sectionInsets = UIEdgeInsets.zero
        
        if ((self.delegate?.collectionView?(self.collectionView!, layout: self, insetForSectionAt: index)) != nil) {
            sectionInsets = (self.delegate?.collectionView!(self.collectionView!, layout: self, insetForSectionAt: index))!
        }
        
        self.sectionInsetses.add(NSValue.init(uiEdgeInsets: sectionInsets))
        
        var interitemSpacing: CGFloat = self.minimumInteritemSpacing
        var lineSpacing: CGFloat = self.minimumLineSpacing
        
        if ((self.delegate?.collectionView?(self.collectionView!, layout: self, minimumInteritemSpacingForSectionAt: index)) != nil) {
            interitemSpacing = (self.delegate?.collectionView!(self.collectionView!, layout: self, minimumInteritemSpacingForSectionAt: index))!
        }
        
        if ((self.delegate?.collectionView?(self.collectionView!, layout: self, minimumLineSpacingForSectionAt: index)) != nil) {
            lineSpacing = (self.delegate?.collectionView!(self.collectionView!, layout: self, minimumLineSpacingForSectionAt: index))!
        }
        
        var itemsContentRect: CGRect = CGRect.zero
        itemsContentRect.origin.x = sectionInsets.left
        itemsContentRect.origin.y = headerHeight + sectionInsets.top
        
        let numberOfColumns = self.delegate?.collectionView(self.collectionView!, layout: self, numberOfColumnInSection: index)
        itemsContentRect.size.width = (collectionView?.frame.size.width)! - (sectionInsets.left + sectionInsets.right)
        
        let columnSpace: CGFloat = itemsContentRect.size.width - (interitemSpacing * (CGFloat)(numberOfColumns! - 1))
        let columnWidth: CGFloat = columnSpace / CGFloat.init(integerLiteral:(numberOfColumns! == 0 ? 1 : numberOfColumns!))
        
        let columns = NSMutableArray.init(capacity: numberOfColumns!)
        for _ in 0 ..< numberOfColumns! {
            columns.add(NSMutableArray.init(capacity: 0))
        }
        
        self.columnRectsInSection.add(columns)

        
        for itemIndex in 0 ..< items {
            let itemIndexPath = IndexPath.init(item: itemIndex, section: index)
            let destinationColumnIndex = self.preferredColumnIndexInSection(index)
            let destinationRowInColumn = self.numberOfItemsForColumn(destinationColumnIndex, inSection: index)
            var lastItemInColumnOffsetY = self.lastItemOffsetYForColumn(destinationColumnIndex, inSection: index)
            
            if destinationRowInColumn == 0 {
                lastItemInColumnOffsetY += sectionRect.origin.y
            }
            
            var itemRect: CGRect = CGRect.zero
            itemRect.origin.x = itemsContentRect.origin.x + CGFloat.init(integerLiteral: destinationColumnIndex) * (interitemSpacing + columnWidth)
            itemRect.origin.y = lastItemInColumnOffsetY + (destinationRowInColumn > 0 ? lineSpacing : sectionInsets.top)
            itemRect.size.width = columnWidth
            itemRect.size.height = columnWidth
            
            if ((self.delegate?.collectionView?(self.collectionView!, layout: self, sizeForItemAt: itemIndexPath)) != nil) {
                let itemSize = self.delegate?.collectionView?(self.collectionView!, layout: self, sizeForItemAt: itemIndexPath)
                itemRect.size.height = itemSize!.height
            }
            
            let itemAttributes = UICollectionViewLayoutAttributes.init(forCellWith: itemIndexPath)
            itemAttributes.frame = itemRect
            (self.layoutItemAttributes.object(at: index) as AnyObject).add(itemAttributes)
            ((self.columnRectsInSection.object(at: index) as AnyObject).object(at: destinationColumnIndex) as AnyObject).add(NSValue.init(cgRect: itemRect))
        }
        
        itemsContentRect.size.height = self.heightOfItemsInSection(indexPath.section) + sectionInsets.bottom
        
        var footerHeight: CGFloat = 0.0
        if ((self.delegate?.collectionView?(self.collectionView!, layout: self, referenceSizeForFooterInSection: index)) != nil) {
            var footerFrame: CGRect = CGRect.zero
            footerFrame.origin.x = 0.0
            footerFrame.origin.y = sectionRect.origin.y
            
            let footerSize = self.delegate?.collectionView!(self.collectionView!, layout: self, referenceSizeForFooterInSection: index)
            
            footerFrame.size.width = (footerSize?.width)!
            footerFrame.size.height = (footerSize?.height)!
            
            let footerAttributes = UICollectionViewLayoutAttributes.init(forSupplementaryViewOfKind: UICollectionElementKindSectionFooter, with: indexPath)
            
            footerAttributes.frame = footerFrame
            
            footerHeight = footerFrame.size.height
            
            self.footerItemAttributes.add(footerAttributes)
        }
        
        if index > 0 {
            itemsContentRect.size.height -= sectionRect.origin.y
        }
        sectionRect.size.height = itemsContentRect.size.height + footerHeight
        
        self.sectionRects.add(NSValue.init(cgRect: sectionRect))
    }
    
    fileprivate func heightOfItemsInSection(_ index: Int) -> CGFloat {
        var maxHeightBetweenColumns : CGFloat = 0.0
        let columnsInSection = self.columnRectsInSection.object(at: index) as! NSArray
        for columnIndex in 0 ..< columnsInSection.count {
            let heightOfColumn = self.lastItemOffsetYForColumn(columnIndex, inSection: index)
            maxHeightBetweenColumns = max(maxHeightBetweenColumns, heightOfColumn)
        }
        return maxHeightBetweenColumns
    }
    
    fileprivate func numberOfItemsForColumn(_ columnIndex: Int, inSection sectionIndex: Int) -> Int {
        let sectionColumns = self.columnRectsInSection.object(at: sectionIndex) as! NSArray
        return (sectionColumns.object(at: columnIndex) as! NSArray).count
    }
    
    fileprivate func lastItemOffsetYForColumn(_ columnIndex: Int, inSection sectionIndex: Int) -> CGFloat {
        let columnsInSection = self.columnRectsInSection.object(at: sectionIndex) as! NSArray
        let itemsInColumn = columnsInSection.object(at: columnIndex)
        if (itemsInColumn as AnyObject).count == 0 {
            if self.headerItemAttributes.count > sectionIndex {
                let headerAttributes = self.headerItemAttributes.object(at: sectionIndex) as! UICollectionViewLayoutAttributes
                let headerFrame = headerAttributes.frame
                return headerFrame.size.height
            }
            return 0
        } else {
            let lastItemRectValue = (itemsInColumn as AnyObject).lastObject as! NSValue
            let lastItemRect = lastItemRectValue.cgRectValue
            return lastItemRect.maxY
        }
    }
    
    fileprivate func preferredColumnIndexInSection(_ index: Int) -> Int {
        var shortestColumnIndex = 0
        var heightOfShortestColumn = CGFloat.greatestFiniteMagnitude
        let columnRects = self.columnRectsInSection.object(at: index) as! NSArray
        for columnIndex in 0 ..< columnRects.count {
            let columnHeight = self.lastItemOffsetYForColumn(columnIndex, inSection: index)
            if columnHeight < heightOfShortestColumn {
                heightOfShortestColumn = columnHeight
                shortestColumnIndex = columnIndex
            }
        }
        return shortestColumnIndex
    }
    
    fileprivate func rectForSectionAtIndex(_ index: Int) -> CGRect {
        if index < 0 || index >= self.sectionRects.count {
            return CGRect.zero
        }
        
        let sectionRectValue = self.sectionRects.object(at: index) as! NSValue
        let sectionRect = sectionRectValue.cgRectValue
        return sectionRect
    }
    
    fileprivate func searchVisibleLayoutAttributesInRect(_ rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        let itemAttrs = NSMutableArray()
        let visibleSection = self.sectionIndexesInRect(rect)
        (visibleSection as NSIndexSet).enumerate({ (sectionIndex: Int, pointer: UnsafeMutablePointer<ObjCBool>) -> Void in
            
            let sectionLayoutItemAttributes = self.layoutItemAttributes.object(at: sectionIndex)
            for i in 0 ..< (sectionLayoutItemAttributes as AnyObject).count {
                let layoutItemAttributes = (sectionLayoutItemAttributes as AnyObject).object(at: i) as! UICollectionViewLayoutAttributes
                layoutItemAttributes.zIndex = 1
                let itemRect = layoutItemAttributes.frame
                let isVisible = rect.intersects(itemRect)
                if isVisible {
                    itemAttrs.add(layoutItemAttributes)
                }
            }
            
            if self.footerItemAttributes.count > sectionIndex {
                let footerLayoutAttributes = self.footerItemAttributes.object(at: sectionIndex) as! UICollectionViewLayoutAttributes
                let isVisible = rect.intersects(footerLayoutAttributes.frame)
                if isVisible {
                    itemAttrs.add(footerLayoutAttributes)
                }
                self.currentEdgeInsets = UIEdgeInsets.zero
            } else {
                let insetsValue = self.sectionInsetses.object(at: sectionIndex) as! NSValue
                self.currentEdgeInsets = insetsValue.uiEdgeInsetsValue
            }
            
            if self.headerItemAttributes.count > sectionIndex {
                let headerLayoutAttributes = self.headerItemAttributes.object(at: sectionIndex) as! UICollectionViewLayoutAttributes
                if self.enableStickyHeaders {
                    let lastItemAttributes = itemAttrs.lastObject as! UICollectionViewLayoutAttributes
                    
                    itemAttrs.add(headerLayoutAttributes)
                    
                    if lastItemAttributes.representedElementKind != UICollectionElementKindSectionHeader {
                        self.updateTheHeaderLayoutAttributes(headerLayoutAttributes, lastItemAttributes: lastItemAttributes)
                    }
                    
                } else {
                    let isVisible = rect.intersects(headerLayoutAttributes.frame)
                    if isVisible {
                        itemAttrs.add(headerLayoutAttributes)
                    }
                }
            }
        })
        return NSArray(array: itemAttrs) as? [UICollectionViewLayoutAttributes]
    }
    
    fileprivate func sectionIndexesInRect(_ rect: CGRect) -> IndexSet {
        let visibleIndexes = NSMutableIndexSet()
        let numberOfSections = self.collectionView?.numberOfSections
        for index in 0 ..< numberOfSections! {
            let sectionRectValue = self.sectionRects.object(at: index) as! NSValue
            let sectionRect = sectionRectValue.cgRectValue
            let isVisible = rect.intersects(sectionRect)
            if isVisible {
                visibleIndexes.add(index)
            }
        }
        return visibleIndexes as IndexSet
    }
    
    fileprivate func updateTheHeaderLayoutAttributes(_ headerAttributes: UICollectionViewLayoutAttributes, lastItemAttributes: UICollectionViewLayoutAttributes) -> Void {
        let currentBounds = self.collectionView?.bounds
        headerAttributes.zIndex = 1024
        headerAttributes.isHidden = false
        
        var origin = headerAttributes.frame.origin
        let sectionMaxY = lastItemAttributes.frame.maxY - lastItemAttributes.frame.size.height + self.currentEdgeInsets.bottom
        let y = (currentBounds?.maxY)! - (currentBounds?.size.height)! + (self.collectionView?.contentInset.top)!
        
        let maxY = min(max(y, headerAttributes.frame.origin.y), sectionMaxY)
        
        origin.y = maxY
        
        headerAttributes.frame = CGRect(x: origin.x, y: origin.y, width: headerAttributes.frame.size.width, height: headerAttributes.frame.size.height)
        
    }
}


